text/0000-semver-for-ts.md

```diff
@@ -0,0 +1,870 @@
+
+[optional-argument]: https://www.typescriptlang.org/play/#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXwAc4A3XZAZwAooAuecjGLVAcwBp4AjO1ZAW04gYASjrEcWYAG4AUKEiwEKdNjzxkBYFAwhg1OgyasOnAPw9+gkWInSZMiCAzwo8ALyESZKgHICODG0cHw4ARgAmYVlHZ053dU1tXUo-AKCQ+AiooA
+
+
+#### Bug fixes
+
+As with runtime code, types may have bugs. We define a ‘bug' here as a mismatch between types and runtime code. That is: if the types allow code which will cause a runtime type error, or if they forbid code which is allowed at runtime, the types are buggy. Types may be buggy by being inappropriately *wider* or *narrower* than runtime.
+
+For example (noting that this list is illustrative, not exhaustive):
+
+-   If a function is typed as accepting `any` but actually requires a `string`, this will cause an error at runtime, and is a bug.
+
+-   If a function is typed as returning `string | number` but always returns `string`, this is a bug. It will not cause an error at runtime, since consumers must "narrow" the type to use it, and narrowing the type would not even be a breaking change. However, the type is incorrect, and it *will* require end users to do unnecessary work.
+
+-   If an interface is defined as having a property which is *not* part of the public API of the runtime object, or if an interface is defined as *missing* a a property which the public API of the runtime object does have, this is a bug.
+
+As with runtime bugs, authors are free to fix type bugs in a patch release. As with runtime code, this may break consumers who were relying on the buggy behavior. However, as with runtime bugs, this is well-understood to be part of the sociotechnical contract of semantic versioning.
+
+In practice, this suggests two key considerations around type bugs:
+
+1.  It is essential that types be well-tested! See discussion below under [**Tooling**](#tooling).
+
+2.  If a given types bug has existed for long enough, an author may choose to treat it as ["intimate API"][intimate] and change the *runtime* behavior to match the types rather than vice versa.
+
+[intimate]: https://twitter.com/wycats/status/918644693759488005
+
+
+#### Strictness
+
+Type-checking in TypeScript behaves differently under different “strictness” settings, and the compiler adds more strictness settings over time. Changes to types which are not breaking under looser compiler settings may be breaking under stricter compiler settings.
+
+For example: a package with `strictNullChecks: false` could make a function return type nullable without the compiler reporting it within the package or the package’s type tests. However, as described above, this is a breaking change for consumers which have `strictNullChecks: true`. (By contrast, a *consumer* may disable strictness settings safely: code which type-checks under a stricter setting also type-checks under a less strict setting.)
+
+Accordingly, conforming packages must use `strict: true` in their compiler settings.
```

Comment:
Given the way the semantics of these work, I'm scratching my head to come up with an example of how having these on vs. off would change a declaration in a way that would matter to downstream consumers; I think the TS team may be treating them as a lint for precisely this reason. While they're effectively exposing `| undefined` in more places, they're doing so in a way that is true *regardless* of whether the consumer has them enabled, and which only matters given `strictNullChecks: true`. That is: if I publish `function foo(): Array<string>`, then *whether or not* I as the author have `noUncheckedIndexedAccess` enabled, the semantics are the same for downstream consumers: dependent on *their* setting.

I think the only way around that would be to publish types as `SafeArray<string>` where

```ts
type SafeArray<T> = Array<T | undefined>;
```

🤔 

It's possible I've missed something, though; I'm going to integrate the above as the text of an unanswered question.

Reply:
In a DM, @dfreeman points out that not checking `noPropertyAccessFromIndexSignature` would cause breakage for consumers with this change:

```diff
- function foo(): { a: string };
+ function foo(): { [key: string]: string };
```

As a consumer:

```ts
foo().a.length;
```

After the diff, this would stop type-checking.

Reply:
Integrated our Discord DM discussion into 824548e.